File: src/commonMain/kotlin/com/xebia/functional/chains/LLMChain.kt
package com.xebia.functional.chains

import arrow.core.Either
import arrow.core.raise.either
import com.xebia.functional.llm.openai.CompletionChoice
import com.xebia.functional.llm.openai.CompletionRequest
import com.xebia.functional.llm.openai.OpenAIClient
import com.xebia.functional.prompt.PromptTemplate

@Suppress("LongParameterList")
suspend fun LLMChain(
    llm: OpenAIClient,
    promptTemplate: PromptTemplate,
    llmModel: String = "text-davinci-003",
    user: String = "testing",
    echo: Boolean = false,
    n: Int = 1,
    temperature: Double = 0.0,
    outputVariable: String,
    chainOutput: Chain.ChainOutput = Chain.ChainOutput.OnlyOutput
): Chain = object : Chain {

    private val inputKeys: Set<String> = promptTemplate.inputKeys.toSet()
    private val outputKeys: Set<String> = setOf(outputVariable)

    override val config: Chain.Config = Chain.Config(inputKeys, outputKeys, chainOutput)

    override suspend fun call(inputs: Map<String, String>): Either<Chain.InvalidInputs, Map<String, String>> =
        either {
            val prompt = promptTemplate.format(inputs)

            val request = CompletionRequest(
                model = llmModel,
                user = user,
                prompt = prompt,
                echo = echo,
                n = n,
                temperature = temperature,
            )

            val completions = llm.createCompletion(request)
            formatOutput(completions)
        }

    private fun formatOutput(completions: List<CompletionChoice>): Map<String, String> =
        config.outputKeys.associateWith {
            completions.joinToString(", ") { it.text }
        }
}

File: scala/src/main/scala/com/xebia/functional/chains/LLMChain.scala
package com.xebia.functional.scala.chains

import cats.effect.Sync
import cats.syntax.all.*

import com.xebia.functional.scala.chains.models.Config
import com.xebia.functional.scala.config.OpenAIConfig
import com.xebia.functional.scala.config.*
import com.xebia.functional.scala.llm.*
import com.xebia.functional.scala.llm.models.OpenAIRequest
import com.xebia.functional.scala.llm.models.*
import com.xebia.functional.scala.llm.openai.OpenAIClient
import com.xebia.functional.scala.prompt.PromptTemplate
import eu.timepit.refined.types.string.NonEmptyString

class LLMChain[F[_]: Sync](
    llm: LLM[F],
    promptTemplate: PromptTemplate[F],
    outputVariable: NonEmptyString,
    onlyOutput: Boolean
) extends BaseChain[F]:
  val config = Config(promptTemplate.inputKeys.toSet, Set(outputVariable.value), onlyOutput)

  def call(inputs: Map[String, String]): F[Map[String, String]] =
    for
      prp <- preparePrompt(inputs)
      cmp <- llm.generateFromPrompt(prp)
      out = formatOutput(cmp)
    yield out

  def preparePrompt(inputs: Map[String, String]): F[String] = promptTemplate.format(inputs)
  def formatOutput(completions: List[LLMResult]): Map[String, String] =
    Map(outputVariable.value -> completions.map(_.generatedText).mkString(", "))

object LLMChain:
  def make[F[_]: Sync](
      llm: LLM[F],
      promptTemplate: PromptTemplate[F],
      outputVariable: NonEmptyString,
      onlyOutput: Boolean
  ): LLMChain[F] =
    new LLMChain[F](llm, promptTemplate, outputVariable, onlyOutput)
